crates/relayer/src/chain/cosmos.rs

```diff
@@ -1933,31 +1933,118 @@ impl ChainEndpoint for CosmosSdkChain {
+            let mut page_key = Vec::new();
 
-        let request = tonic::Request::new(request.into());
+            let mut limit = request.pagination.get_limit();
 
-        let response = self
-            .block_on(client.packet_commitments(request))
-            .map_err(|e| Error::grpc_status(e, "query_packet_commitments".to_owned()))?
-            .into_inner();
+            loop {
+                crate::time!(
+                    "query_packet_commitments_loop_iteration",
+                    {
+                        "src_chain": self.config().id.to_string(),
+                    }
+                );
+                let mut raw_request =
+                    ibc_proto::ibc::core::channel::v1::QueryPacketCommitmentsRequest::from(
+                        request.clone(),
+                    );
 
-        let mut commitment_sequences: Vec<Sequence> = response
-            .commitments
-            .into_iter()
-            .map(|v| v.sequence.into())
-            .collect();
-        commitment_sequences.sort_unstable();
+                if let Some(pagination) = raw_request.pagination.as_mut() {
+                    pagination.key = page_key;
+                }
 
-        let height = response
-            .height
-            .and_then(|raw_height| raw_height.try_into().ok())
-            .ok_or_else(|| Error::grpc_response_param("height".to_string()))?;
+                let mut tonic_request = tonic::Request::new(raw_request);
+                tonic_request.set_timeout(Duration::from_secs(10));
+
+                let response = self.rt.block_on(async {
+                    client
+                        .packet_commitments(tonic_request)
+                        .await
+                        .map_err(|e| Error::grpc_status(e, "query_packet_commitments".to_owned()))
+                });
+
+                match response {
+                    Ok(response) => {
+                        let inner_response = response.into_inner().clone();
+                        let next_key = inner_response
+                            .pagination
+                            .as_ref()
+                            .map(|p| p.next_key.clone());
+
+                        results.push(Ok(inner_response));
+
+                        match next_key {
+                            Some(next_key) if !next_key.is_empty() => {
+                                page_key = next_key;
+                            }
+                            _ => break,
+                        }
+                    }
+                    Err(e) => {
+                        results.push(Err(e));
+                        break;
+                    }
+                }
+                if limit == 0 {
+                    break;
+                }
+                limit -= 1;
+            }
+
+            let responses =
+                results.into_iter().collect::<Result<
+                    Vec<ibc_proto::ibc::core::channel::v1::QueryPacketCommitmentsResponse>,
```

Comment:
Can't the inner type be elided here?

```suggestion
                    Vec<_>,
```